Source/Core/Core/AchievementManager.cpp

```diff
@@ -544,13 +563,41 @@ AchievementManager::ResponseType AchievementManager::StartRASession()
 
 AchievementManager::ResponseType AchievementManager::FetchGameData()
 {
-  std::string username = Config::Get(Config::RA_USERNAME);
-  std::string api_token = Config::Get(Config::RA_API_TOKEN);
-  rc_api_fetch_game_data_request_t fetch_data_request = {
-      .username = username.c_str(), .api_token = api_token.c_str(), .game_id = m_game_id};
-  return Request<rc_api_fetch_game_data_request_t, rc_api_fetch_game_data_response_t>(
-      fetch_data_request, &m_game_data, rc_api_init_fetch_game_data_request,
-      rc_api_process_fetch_game_data_response);
+  rc_api_fetch_game_data_request_t fetch_data_request;
+  std::string username, api_token;
+  u32 game_id;
+  {
+    std::lock_guard lg{m_lock};
+    username = Config::Get(Config::RA_USERNAME);
+    api_token = Config::Get(Config::RA_API_TOKEN);
+    game_id = m_game_id;
+  }
+  fetch_data_request = {
+      .username = username.c_str(), .api_token = api_token.c_str(), .game_id = game_id};
+  ResponseType r_type =
+      Request<rc_api_fetch_game_data_request_t, rc_api_fetch_game_data_response_t>(
+          fetch_data_request, &m_game_data, rc_api_init_fetch_game_data_request,
+          rc_api_process_fetch_game_data_response);
+  if (r_type == ResponseType::SUCCESS)
+  {
+    std::lock_guard lg{m_lock};
+    if (game_id != m_game_id)
+    {
+      INFO_LOG_FMT(ACHIEVEMENTS,
+                   "Attempted to retrieve game data for ID {}; running game is now ID {}", game_id,
+                   m_game_id);
+      rc_api_destroy_fetch_game_data_response(&m_game_data);
+      return ResponseType::EXPIRED_CONTEXT;
+    }
+    INFO_LOG_FMT(ACHIEVEMENTS, "Retrieved game data for ID {}.", game_id);
+  }
+  else
+  {
+    WARN_LOG_FMT(ACHIEVEMENTS, "Failed retrieving game data for ID {}.", game_id);
+    std::lock_guard lg{m_lock};
+    rc_api_destroy_fetch_game_data_response(&m_game_data);
```

Comment:
Is this right...? The [documentation](https://github.com/RetroAchievements/rcheevos/wiki/rc_api_process_fetch_game_data_response) is somewhat unclear on when the response counts as 'constructed' -- only on success of `rc_api_process_fetch_game_data_response()`? Always?

Reply:
I went through the code and what I'm seeing is that _process_ doesn't roll back anything if it fails, and _destroy_ doesn't seem to fail if anything is missing. So from what I can tell, if the object is partially created, I'm better off destroying to make sure it's completely clean for the next use and isn't leaking.